genesis-tools/lockup-contract/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+//! A smart contract that allows lockup of a smart contract.
+use borsh::{BorshDeserialize, BorshSerialize};
+use key_management::{KeyType, PublicKey};
+use near_bindgen::{env, near_bindgen, Promise};
+use std::collections::{HashMap, HashSet};
+
+mod key_management;
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+#[near_bindgen]
+#[derive(Default, BorshDeserialize, BorshSerialize)]
+pub struct LockupContract {
+    lockup_amount: u128,
+    lockup_timestamp: u64,
```

Comment:
We should not rely on timestamp because otherwise validators are incentivized to push timestamp forward as much as possible when they produce blocks. Instead we should use block index for vesting.

Reply:
They can only push the timestamp forward as much as our threshold that invalidates incorrect timestamps allows, which is going to be less than half a minute in practice :) On the other hand, we have no real idea what is our block production speed is going to be, while we have very strict vesting and lockup schedules already predefined in the paperwork. What if we start producing our blocks 50% slower or faster in the next year after the launch are we going to unlock/vest the tokens 2 years earlier/later? This looks like a much nastier consequence than potentially having an incentive that will hypothetically cause validators lie in the block production by several seconds, maybe.